### DISCUSSION TITLE
UHUGEINT

### DISCUSSION THREAD

**Discussion by valmunos** [duckdb/duckdb#2945] [Ideas]
Curious when the data type UHUGEINT will be supported?  I find it helpful for dealing with IPV6 addresses.

**Comment by hannes** [comment]
Quick questions, IPv6 addresses are 64 bits, so would `UBIGINT` not suffice?

**Comment by valmunos** [comment]
According to [this Wikipedia article](https://en.wikipedia.org/wiki/IPv6):

"IPv6 uses 128-bit addresses, theoretically allowing 2^128, or approximately 3.4×10^38 total addresses."

**Comment by valmunos** [comment]
I think I have found a way to work with them, but it is a cumbersome 2-step process.

- First, separate all IPv6 <= 2 ^127 - 1.  Those can be handled without any adjustments.
- For the remaining IPs, subtract a constant from them such that the largest one is equal to 2 ^ 127 - 1.

If you are joining these on a CIDR range as I am, you will then need to adjust the range by the same constant as you did in the second step.  Then you can join them as you normally would with the BETWEEN operator.

Also, it turns out that, in practice, most (but not all) values fall below this threshold.

**Comment by hannes** [comment]
Oh my mistake, sorry.

**Comment by digizeph** [comment]
I'd like to bump up this discussion. I wonder if there is any techincal blockers to support UHUGEINT? having the ability to store IPv6 really unblocks DuckDB's potential in network analytics and applications.

**Comment by lnkuiper** [comment]
Maybe this can be added to our INET extension instead? It seems like there is some support for IPv6 code, but it's incomplete. I think this would be preferred over adding the UHUGEINT type.

**Comment by ak-coram** [comment]
I think UUIDs can also be abused to represent them:

```lisp
(ddb:get-result
 (ddb:q "SELECT ?::uuid AS address"
        (fuuid:from-integer 
         (parse-integer (remove #\: "2001:db8:85a3:8d3:1319:8a2e:370:7348")
                        :radix 16)))
 'address 0)
;; => #<FRUGAL-UUID:UUID 0002001d-b885-a38d-3131-98a2e3707348>
```

**Comment by troycurtisjr** [comment]
Is anyone known to be working on adding ipv6 to the inet extension already?
